### DISCUSSION TITLE
[Q&A] FQDN of the client

### DISCUSSION THREAD

**Discussion by liedekef** [glpi-project/glpi-agent#345] [Q&A]
### Documentation & bug reporting acknowledgment

Yes, I read it

### Describe your problem

When registering a server via an agent, an asset gets created but with the short hostname. Is there a way to either use the fqdn or somehow track the dns-info from the host? I do see the info present in the xml:
```
      <DNS_DOMAIN>xxxx.be</DNS_DOMAIN>
      <FQDN>laptop-franky.xxxx.be</FQDN>
```
so it is being sent :-)

**Comment by g-bougard** [comment]
Hi @liedekef 
as far I know, `FQDN` & `DNS_DOMAIN` are not used during automatic inventory.
You should probably open a github issue on glpi project and request a feature to have an option to use fqdn as asset name in place of shortname, when possible.

**Comment by liedekef** [comment]
Well, it seems glpi uses hardware=>name as name for the asset. So: for network-discovered assets that's (most of the time) set to the dns name, but for agent-created inventories it is in fact the "name" of the device.
I could think of some possibilities here:
- allow a config param to set the name of the device
- use the fqdn as name

My preference would go to a config param. Reason behind this: imagine a laptop that connects from work (having a work-domain related fqn) and then later on connects from home (where, before vpn is active, the fqdn might be related to your internal domain, or the domain of your isp). Allowing to set the name of the device in the config would allow for a more consistent behavior.

Next to that, it seems that for linux the short hostname is always "forced":
./lib/GLPI/Agent/Task/Inventory/Generic/Hostname.pm:    my $hostname = getHostname(short => 1);
==> since that file is actually used to set the hardware name alone, one might reconsider the "Hostname.pm" name :-)
==> changing that call to remove "short=>1" results in the hostname of the laptop to be taken (in case it is fqdn, it will not get cut off then). While I still think a config param might help, this is already a good simple "fix" I would say

**Reply by g-bougard** [reply]
Hi @liedekef 

we have here `short => 1` as the `getHostname()` API can return fqdn when called from other places in the code. So yes, this API call is "forced". It was decided historically to use the short name here. I think I wasn't on the project when this was decided, so I can't tell you definitively the right reason for that choice. I can only guess the short name choice was due to the fact the fqdn can be really very very long.

Actually, you have a solution to set the computer name to anything you want. This is a little tricky as you'll have to use the `additional-content` option using json format for the inventory as it permits to force the update of one value on any node out of an array (this is the case for the computer name). So you can set a file path to a json file on this parameter. The json file must be set to the right format, in your case, this could be:
```
{
   "content": {
      "hardware": {
         "name": "laptop-franky.xxxx.be"
      }
   }
}
```
You can imagine to create that json file during the installation and leave as is for the computer life or create a scheduled task which generate it one time a day in the case the computer name can change during the computer life.

About a new param, only adding a config param to set the hostname as fqdn in place of short version is a possibility. Here, we can't change the default comportment of the agent or we will have a lot of people who will complain about this change.

Anyway, I would really prefer a GLPI solution:
 1. GLPI need to import FQDN and/or DNS_DOMAIN
 2. GLPI should support:
    1. an option to display computer name as FQDN
    2. or allowing to display the fqdn or dns_domain column

As I said you should probably open an issue on GLPI github project asking for the best solution in your point of view from 2.i or 2.ii. If you do, just link this discussion for the history.

**Comment by liedekef** [comment]
I understand that the additional-content option is possible (and apparently it supports xml and json files) so that is an option, but: would you be willing to consider a PR adding the "agent name" as a config option? This would be more in line with other tooling (like zabbix, nagios, ...) that allow to set the client name if wanted (and take the original behavior if nothing is present, in this case: the short hostname).
Since the agent already uses the fqdn for remotely discovered devices, I don't think the glpi server itself will mind :-)
The other thing (concerning fqdn and/or dns_domain) is indeed something to be fixed in glpi itself.

**Reply by g-bougard** [reply]
Any community contribution will of course be considered for integration, especially if it adds a new feature without breaking anything.
Here you're talking about the "agent name". I'm confused, you were talking before about the "asset name". Is that another discussion ?

**Reply by liedekef** [reply]
No, it is indeed the asset name. But from zabbix/... point of view, that typically corresponds with "agent name" :-)
I'll see if I can cook up something. If so: any preference on the name of the option added? Or is asset_name ok?

**Reply by g-bougard** [reply]
Okay, I understand now.
Not sure an option to set the asset name with an arbitrary value is a good one. This can be dangerous if used with the same value (by error of course) for many computers. An option to set fqdn as asset name would be safer IMHO. And you could name it `use-fqdn-as-assetname` ;-)

**Reply by liedekef** [reply]
or ... `use-hostname-as-assetname`  ? I think maybe to just create an extra option to not use the "short=>1" if that option is chosen, but of course the hostname is not always fqdn to begin with (so "fqdn" is not totally correct).

**Reply by g-bougard** [reply]
You're right.
Maybe `use-full-hostname-as-assetname` is even better as a full hostname still can be short. But this starts to be very long name for a parameter... Maybe a shorter one like `hostname-support` which would default to "1" meaning short name should be used. And if set to "2", this would force to not use the `short => 1` parameter in the call.

**Reply by liedekef** [reply]
If we go for a shorter one: `assetname-support` (default: 1, meaning short hostname, other option 2: meaning hostname-as-is, and maybe other options can be added future-wise) ?

**Reply by g-bougard** [reply]
Nice

**Reply by liedekef** [reply]
PR created :-)

**Reply by g-bougard** [reply]
Yes, seen it: #347
& thank you ;-)

**Reply by g-bougard** [reply]
Hi @liedekef 

in the case you don't know, development packages have been generated as artifacts for this GH Actions run: https://github.com/glpi-project/glpi-agent/actions/runs/4234335856

Have you tested your solution and does it produces what you expect in GLPI ?

**Reply by liedekef** [reply]
Nice (I don't know/use GH actions yet). In my tests (using the param as commandline option to the inventory, it produces the correct XML as expected). When the param is absent (or something else besides "2", the original behavior works).
Using your package, I'll update a test server and try the agent too.

**Reply by liedekef** [reply]
I can confirm that it works as expected for the agent too: setting "assetname-support = 2" in a config file uses the hostname as-is.
For future references: maybe you can add the parsed config values to the debug output. I now just used a "warn" statement to extra verify the values being loaded in my tests, but that's not ok I'd say :-) .
